src/mame/machine/cxd8452aq.cpp

```diff
@@ -0,0 +1,280 @@
+// license:BSD-3-Clause
+// copyright-holders:Brice Onken
+// thanks-to:Patrick Mackinlay
+
+/*
+ * Sony CXD8452AQ WSC-SONIC3 APbus interface and Ethernet controller controller
+ *
+ * TODO:
+ *  - Error-related interrupts
+ */
+
+#include "emu.h"
+#include "cxd8452aq.h"
+
+#define LOG_GENERAL (1U << 0)
+#define LOG_REGISTERS (1U << 1)
+#define LOG_INTERRUPT (1U << 2)
+#define LOG_DATA (1U << 3)
+
+#define SONIC3_DEBUG (LOG_GENERAL | LOG_INTERRUPT)
+#define SONIC3_TRACE (SONIC3_DEBUG | LOG_DATA)
+
+#include "logmacro.h"
+
+DEFINE_DEVICE_TYPE(CXD8452AQ, cxd8452aq_device, "cxd8452aq", "Sony CXD8452AQ WSC-SONIC3")
+
+namespace
+{
+	// control register constants
+	static constexpr uint32_t INT_EN_MASK = 0x7f00;
+
+	// TODO: determine if external interrupt should be included when cleared
+	static constexpr uint32_t INT_CLR_MASK = 0xf0;
+	static constexpr uint32_t RX_DMA_COMPLETE = 0x40;
+	static constexpr uint32_t TX_DMA_COMPLETE = 0x20;
+	static constexpr uint32_t EXT_INT = 0x1;
+
+	// count register constants
+	static constexpr uint32_t DMA_START = 0x80000000;
+
+	// DMA update timer
+	// TODO: Actual frequency, since we don't have DRQ to implictly rate limit
+	//       Might be the APbus frequency.
+	static constexpr int DMA_TIMER = 100;
+}
+
+cxd8452aq_device::cxd8452aq_device(const machine_config &mconfig, const char *tag, device_t *owner, uint32_t clock)
+	: device_t(mconfig, CXD8452AQ, tag, owner, clock),
+	  device_memory_interface(mconfig, *this),
+	  main_bus_config("main_bus", ENDIANNESS_BIG, 32, 32, 0),
+	  sonic_config("sonic", ENDIANNESS_BIG, 32, 32, 0, address_map_constructor(FUNC(cxd8452aq_device::sonic_bus_map), this)),
+	  m_irq_handler(*this),
+	  m_apbus_virt_to_phys_callback(*this),
+	  m_bus(*this, finder_base::DUMMY_TAG, -1, 64) {}
+
+device_memory_interface::space_config_vector cxd8452aq_device::memory_space_config() const
+{
+	// Uses a similar trick that the Jazz MCT-ADR driver uses to translate accesses from the SONIC to the system bus
```

Comment:
I think I am missing something important for this one. The other uses I can find for SONICs all assign them a bus, some of them directly to the CPU bus, 
which I think would cause issues because of the required translation. Is there a standard mechanism for hooking up callbacks to internal `required_address_space`s 
that I should consider using for this? The part I don't fully understand is what I would pass in to the SONIC's `set_bus` method - would this need to be a new 
`required_address_space` owned by the news_r4k driver that links together the DMA controller, host accesses, and SONIC, instead of using a `device_memory_interface`?

(Also, apologies for the late reply, build and competition season for the robotics team I mentor runs from Jan-Apr, so I have had no free time lately!)